---
title: "nucleoSim: Simulate synthetic nucleosome maps"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Simulate synthetic nucleosome maps}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignettePackage{nucleoSim}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```
<br>
**Package**: `r Rpackage("nucleoSim")`<br />
**Authors**: `r packageDescription("nucleoSim")[["Author"]]`<br />
**Version**: `r packageDescription("nucleoSim")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("nucleoSim")[["License"]]`<br />

# Licensing and citing

This package and the underlying `r Rpackage("nucleoSim")` code are 
distributed under the Artistic license 2.0. You are free to use and 
redistribute this software.

If you use this package for a publication, we would ask you to cite the 
following:

> Samb R, Khadraoui K, Belleau P, et al. (2015). "Using informative Multinomial-Dirichlet prior in a t-mixture with reversible jump estimation of nucleosome positions for genome-wide profiling." Statistical Applications in Genetics and Molecular Biology. Volume 14, Issue 6, Pages 517–532, ISSN (Online) 1544-6115, ISSN (Print) 2194-6302, December 2015. <a href="http://dx.doi.org/10.1515/sagmb-2014-0098">doi: 10.1515/sagmb-2014-0098</a>

> Flores O and Orozco M (2011). “nucleR: a package for non-parametric nucleosome positioning.” Bioinformatics, 27, pp. 2149–2150. <a href="http://dx.doi.org/10.1093/bioinformatics/btr345">doi: 10.1093/bioinformatics/btr345</a>


# Introduction

`r Rpackage("nucleoSim")` can simulate datasets for nucleosomes experiments. 
It allows the user to introduce various 'contaminants' into the sequence 
datasets, such as fuzzy nucleosomes and missing nucleosomes, in order to be 
more realistic and to unable the evaluation of the influence of common 
'noice' on the detection of nucleosome.

* Select the number of well-positioned nucleosomes to generate
* Select the number of well-positioned nucleosomes to be randomly deleted
* Select the number of fuzzy nucleosomes to add
* Select the type of distribution used to assign the 
start position to the reads associated to the nucleosomes
* Select the variance associated to the distribution for the well-positioned 
nucleosomes
* Select the variance associated to the distribution for the fuzzy
nucleosomes
* Select the variance associated to the Normal distribution used to assign
the length of each read
* Select the maximum read coverage associated to one nucleosome
* Select the length of the DNA linker
* Select the length of the nucleosomes (not recommanded to change default value)
* Select a seed when results need to be reproducible


# Loading `r Rpackage("nucleoSim")` package

As with any R package, the `r Rpackage("nucleoSim")` package should first be 
loaded with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(nucleoSim)
```


# Description

The packages can generate 2 types of synthetic data sets:

**Synthetic Nucleosome Maps**: A map with complete reads covering the 
nucleosome regions
    
**Synthetic Nucleosome Samples**: A map with forward and reverse reads 
emulating those obtained using a next-generation sequencing technology on a
nucleosome map 


## Synthetic Nucleosome Maps

A synthetic nucleosome map is a section of genome covered by a fixed number
of nucleosomes. Each nucleosome beeing associated to a specific number of 
reads. The parameters passed to the `syntheticNucMapFromDist()` function are
going to affect the distribution of the nucleosomes, as well as, the reads.

Technically, the synthetic nucleosome map is separated into 3 steps:

**1. Adding well-positioned nucleosomes**

The synthetic nucleosome map is splitted into a fixed number of 
sections (`wp.num`) of  equal length (`(nuc.len + lin.len)` 
bases). The center of the nucleosomes are positioned at a fixed number of bases 
from the beginning of each section (`round(nuc.len/2)` bases). Reads are 
assigned, to each nucleosome, using an uniform distribution. The number of 
reads, assigned to each nucleosome, can vary from 1 to `max.cover`. 
The distribution (`distr`), as 
well as the variance (`wp.var`) are used to add some fluctuation to the 
starting position of the reads, which as a mean position corresponding to 
the sarting position of a region. Some fluctuation ot the length of 
the reads is also added following
a normal distribution with a fixed variance (`len.var`). The mean 
length of the reads correpond to the length of the nucleosomes (`nuc.len`).

**2. Deleting some well-positioned nucleosomes**

Following an uniform distribution, a fixed number of nucleosomes (`wp.del`) 
are deleted. Nucleosomes are considered deleted when all reads associated to it
are eliminated.

**3. Adding fuzzy nucleosomes**

A fixed number of fuzzy nucleosomes (`fuz.num`) are added. The position of the
fuzzy nucleosomes is selected following an uniform distribution. Such as for
the well-positioned nucleosomes, reads are 
assigned, to each fuzzy nucleosome, using an uniform distribution. The number 
of reads assigned can vary from 1 to `max.cover`. 
The distribution (`distr`), as 
well as the variance (`wp.var`) are used to add some fluctuation to the 
starting position of the reads, which as a mean position corresponding to 
the sarting position of a region. Some fluctuation ot the length of 
the reads is also added following
a normal distribution with a fixed variance (`len.var`). The mean 
length of the reads correpond to the length of the nucleosomes (`nuc.len`).

### Create a nucleosome map using syntheticNucMapFromDist()

This is an example showing how a synthetic nucleosome map can be generated.

```{r demoMap, warning=FALSE, message=FALSE, collapse=TRUE}
wp.num           <- 20         ### Number of well-positioned nucleosomes
wp.del           <- 5          ### Number of well-positioned nucleosomes to delete
wp.var           <- 30         ### variance associated to the well-positioned nucleosomes
fuz.num          <- 5          ### Number of fuzzy nucleosomes
fuz.var          <- 50         ### Variance associated to the fuzzy nucleosomes
max.cover        <- 70         ### Maximum reads associated to one nucleosome (default: 100)
nuc.len          <- 147        ### Length of the nucleosome (default: 147)
len.var          <- 12         ### variance associated to the length of the reads (default: 10)
lin.len          <- 20         ### Length of the DNA linker (default: 20)
distr            <- "Normal"   ### Type of distribution to use

rnd.seed         <- 210001     ### Set seed when result need to be reproducible
        
#### Create a synthetic nucleosome map
nucleosomeMap <- syntheticNucMapFromDist(wp.num=wp.num, wp.del=wp.del, wp.var=wp.var, 
                                    fuz.num=fuz.num, fuz.var=fuz.var, max.cover=max.cover,
                                    nuc.len=nuc.len, len.var=len.var, lin.len=lin.len, 
                                    rnd.seed=rnd.seed, distr=distr)

#### The start positions of all well-positioned nucleosomes
nucleosomeMap$wp.starts

#### The number of reads associated to each well-positioned nucleosome
nucleosomeMap$wp.nreads

#### A IRanges object containing all reads for the well-positioned nucleosomes
head(nucleosomeMap$wp.reads, n = 2)

#### The start positions of all fuzzy nucleosomes
nucleosomeMap$fuz.starts

#### The number of reads associated to each fuzzy nucleosome
nucleosomeMap$fuz.nreads

#### A IRanges object containing all reads for the fuzzy nucleosomes
head(nucleosomeMap$fuz.reads, n = 2)

#### A IRanges object containing all reads from well-potitioned and fuzzy nucleosomes
head(nucleosomeMap$syn.reads, n = 2)
```

The synthetic nucleosome map can easily be visualised using `plot()` function.
On the graph, each nucleosome is located on the graph using the 
coordonnates:

`   (x,y) = 
(the central position of the nucleosome, the number of reads associated to
the nucleosome)`

```{r showMap, fig.align='center', fig.height=5, fig.width=8}
#### Create visual representation of the synthetic nucleosome map
plot(nucleosomeMap, xlab="Position", ylab="Coverage")
```

## Synthetic Nucleosome Samples

A synthetic nucleosome sample is a map with forward and reverse reads 
emulating those obtained using a next-generation sequencing technology. It
is created using the same 3 first steps than the 
synthetic nucleosome map. However, some new steps are presents:

**1. Adding well-positioned nucleosomes **

The synthetic nucleosome map is splitted into a fixed number of 
sections (`wp.num`) of  equal length (`(nuc.len + lin.len)` 
bases). The center of the nucleosomes are positioned at a fixed number of bases 
from the beginning of each section (`round(nuc.len/2)` bases). Reads are 
assigned, to each nucleosome, using an uniform distribution. The number of 
reads, assigned to each nucleosome, can vary from 1 to `max.cover`. 
The distribution (`distr`), as 
well as the variance (`wp.var`) are used to add some fluctuation to the 
starting position of the reads, which as a mean position corresponding to 
the sarting position of a region. Some fluctuation ot the length of 
the reads is also added following
a normal distribution with a fixed variance (`len.var`). The mean 
length of the reads correpond to the length of the nucleosomes (`nuc.len`).

**2. Deleting some well-positioned nucleosomes**

Following an uniform distribution, a fixed number of nucleosomes (`wp.del`) 
are deleted. Nucleosomes are considered deleted when all reads associated to it
are eliminated.

**3. Adding fuzzy nucleosomes**

A fixed number of fuzzy nucleosomes (`fuz.num`) are added. The position of the
fuzzy nucleosomes is selected following an uniform distribution. Such as for
the well-positioned nucleosomes, reads are 
assigned, to each fuzzy nucleosome, using an uniform distribution. The number 
of reads assigned can vary from 1 to `max.cover`. 
The distribution (`distr`), as 
well as the variance (`wp.var`) are used to add some fluctuation to the 
starting position of the reads, which as a mean position corresponding to 
the sarting position of a region. Some fluctuation ot the length of 
the reads is also added following
a normal distribution with a fixed variance (`len.var`). The mean 
length of the reads correpond to the length of the nucleosomes (`nuc.len`).

**4. Adding an offset**

An offset (`offset`) is added to all nucleosomes and
reads positions to ensure that all values are positives (mainly pertinent for
reads).

**5. Splitting reads into forward and reverse reads**

The reads are splitted into forward and reverse reads of fixed length
(`read.len`). Both ends of the reads are used as starting positions for the
paired-end reads.

The forward and reverse reads are stored in a `data.frame` containing the name 
of a synthetic chromosome (`chr_SYNTHETIC`), the starting position, the ending
position and the direction of all reads.


### Create a nucleosome sample using syntheticNucReadsFromDist()

This funciton needs information about the nucleosomes and their distribution to 
generate a nucleosome sample. The output is of class `syntheticNucMap`.

```{r demoSample, warning=FALSE, message=FALSE, collapse=TRUE}
wp.num           <- 30            ### Number of well-positioned nucleosomes
wp.del           <- 10            ### Number of well-positioned nucleosomes to delete
wp.var           <- 30            ### variance associated to the well-positioned nucleosomes
fuz.num          <- 10            ### Number of fuzzy nucleosomes
fuz.var          <- 50            ### Variance associated to the fuzzy nucleosomes
max.cover        <- 90            ### Maximum reads associated to one nucleosome (default: 100)
nuc.len          <- 147           ### Length of the nucleosome (default: 147)
len.var          <- 12            ### variance associated to the length of the reads (default: 10)
lin.len          <- 20            ### Length of the DNA linker (default: 20)
read.len         <- 45            ### Length of the generated forward and reverse reads (default: 40)
distr            <- "Uniform"     ### Type of distribution to use
offset           <- 10000         ### The number of bases used to offset all nucleosomes and reads

rnd.seed         <- 202           ### Set seed when result need to be reproducible
        
#### Create Uniform sample
nucleosomeSample <- syntheticNucReadsFromDist(wp.num=wp.num, wp.del=wp.del, wp.var=wp.var,
                                        fuz.num=fuz.num, fuz.var=fuz.var, max.cover=max.cover,
                                        nuc.len=nuc.len, len.var=len.var, read.len=read.len, 
                                        lin.len=lin.len, rnd.seed=rnd.seed, distr=distr, 
                                        offset=offset)

#### The central position of all well-positioned nucleosomes with the
#### number of paired-end reads each associated to each one
head(nucleosomeSample$wp, n = 2)

#### The central position of all fuzzy nucleosomes with the
#### number of paired-end reads each associated to each one
head(nucleosomeSample$fuz, n = 2)

#### A data.frame with the name of the synthetic chromosome, the starting 
#### position, the ending position and the direction of all forward and
#### reverse reads
head(nucleosomeSample$dataIP, n = 2)
```

The synthetic nucleosome sample can easily be visualised using `plot()` 
function. On the graph, each nucleosome is located on the graph using the 
coordonnates:

`   (x,y) = 
(the central position of the nucleosome, the number of reads associated to
the nucleosome)`

```{r showSample, fig.align='center', fig.height=5, fig.width=8}
#### Create visual representation of the synthetic nucleosome sample
plot(nucleosomeSample, xlab="Position", ylab="Coverage (number of reads)")
```

### Create a nucleosome sample using syntheticNucReadsFromMap()

A synthetic nucleosome sample can be created using a nucleosome map. The 
nucleosomes and reads present in the nucleosome map will be added an offset. 
Forward and reverse reads will also be generated. The output is of class
`syntheticNucMap`.

```{r demoSampleFromMap, warning=FALSE, message=FALSE, collapse=TRUE}

#### A nucleosome map has already been created 
class(nucleosomeMap)

#### 
read.len         <- 45   ### The length of the reverse and forward reads
offset           <- 500  ### The number of bases used to offset all nucleosomes and reads

#### Create nucleosome sample
nucleosomeSampleFromMap <- syntheticNucReadsFromMap(nucleosomeMap, read.len=read.len, 
                                                        offset=offset)

#### A data.frame with the name of the synthetic chromosome, the starting 
#### position, the ending position and the direction of all forward and
#### reverse reads
head(nucleosomeSampleFromMap$dataIP, n = 2)
```


# Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
