---
title: "nucleoSim: Simulate synthetic nucleosome maps"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Simulate synthetic nucleosome maps}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignettePackage{nucleoSim}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```
<br>
**Package**: `r Rpackage("nucleoSim")`<br />
**Authors**: `r packageDescription("nucleoSim")[["Author"]]`<br />
**Version**: `r packageDescription("nucleoSim")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("nucleoSim")[["License"]]`<br />

## Licensing and citing

This package and the underlying `r Rpackage("nucleoSim")` code are 
distributed under the Artistic license 2.0. You are free to use and 
redistribute this software.

If you use this package for a publication, we would ask you to cite the following:

> Samb R, Khadraoui K, Belleau P, et al. (2015). "Using informative Multinomial-Dirichlet prior in a t-mixture with reversible jump estimation of nucleosome positions for genome-wide profiling." Statistical Applications in Genetics and Molecular Biology. Volume 14, Issue 6, Pages 517–532, ISSN (Online) 1544-6115, ISSN (Print) 2194-6302, December 2015. <a href="http://dx.doi.org/10.1515/sagmb-2014-0098">doi: 10.1515/sagmb-2014-0098</a>

> Flores O and Orozco M (2011). “nucleR: a package for non-parametric nucleosome positioning.” Bioinformatics, 27, pp. 2149–2150. <a href="http://dx.doi.org/10.1093/bioinformatics/btr345">doi: 10.1093/bioinformatics/btr345</a>

## Introduction

`r Rpackage("nucleoSim")` can simulate datasets for nucleosomes experiments. 
It allows the user to introduce various 'contaminants' into the sequence 
datasets, such as fuzzy nucleosomes and missing nucleosomes, in order to be 
more realistic and to unable the evaluation of the influence of common 
'noice' on nucleosomes detection.

## Loading nucleoSim package

As with any R package, the `r Rpackage("nucleoSim")` package should first be 
loaded with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(nucleoSim)
```

## Description

syntheticNucMapFromDist

A synthetic genome is splitted in `wp.num` sections of `(nuc.len + lin.len)` based. 
The nucleosomes are positioned at `round(nuc.len/2)` bases from the beginning of each section.
A number of reads is assigned to each
nucleosome using a Uniform distribution. The number of reads can vary from 
1 to `max.cover`. The defaut start position of all reads is set to the first 
position of each section. However, the distribution selected by the user, as well
as the `wp.var` are used to add some variation to the starting position of the reads.




## Demo


```{r demo, warning=FALSE, message=FALSE, collapse=TRUE}
val.num       <- 130  ### nombre de nucleosomes non-chevauchants
val.del       <- 30
val.var       <- 30   ### variance si autre dist
val.fuz.var   <- 50   ### variance si sautre
val.max.cover <- 70
val.nuc.len   <- 147
val.len.var   <- 10
val.lin.len   <- 20
val.offset    <- 10000

val.rnd.seed <- 202
        
## Create Uniform sample
sample <- syntheticNucReadsFromDist(wp.num=val.num,
                                            wp.del=val.del,
                                            wp.var=val.var,
                                            fuz.num=2*val.del,
                                            fuz.var=val.fuz.var,
                                            max.cover=val.max.cover,
                                            nuc.len=val.nuc.len,
                                            len.var=val.len.var,
                                            lin.len=val.lin.len,
                                            rnd.seed=val.rnd.seed,
                                            distr="Uniform",
                                            offset=val.offset)

head(sample$wp)

head(sample$fuz)
```

```{r showSample, fig.align='center', fig.height=5, fig.width=8}
plot(sample, xlab="Position", ylab="Coverage")
```


## Session info

Here is the output of `sessionInfo()` on the system on which this document was 
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
